c/tskit/trees.c

```diff
@@ -2636,12 +2639,568 @@ check_sites(const tsk_id_t *sites, tsk_size_t num_sites, tsk_size_t num_site_row
+            if (in_parent != NULL) {
+                for (k = 0; k < state_dim; k++) {
+                    tsk_bit_array_get_row(
+                        r_samples, (state_dim * (tsk_size_t) c) + k, &samples_row);
+                    tsk_bit_array_get_row(&child_samples, k, &child_samples_row);
+                    tsk_bit_array_subtract(&samples_row, &child_samples_row);
+                }
+            }
+            in_parent = &child_samples;
+            c = p;
+            p = r_state->parent[p];
+        }
+        for (k = 0; k < state_dim; k++) {
+            tsk_bit_array_get_row(
+                r_samples, (state_dim * (tsk_size_t) c) + k, &samples_row);
+            tsk_bit_array_get_row(&child_samples, k, &child_samples_row);
+            tsk_bit_array_subtract(&samples_row, &child_samples_row);
+        }
+        c = edges_child[e];
+        r_state->branch_len[c] = 0;
+        r_state->parent[c] = TSK_NULL;
+    }
+    for (j = 0; j < r_state->n_edges_in; j++) {
+        e = r_state->edges_in[j];
+        c = edges_child[e];
+        p = edges_parent[e];
+        tsk_memset(child_samples.data, 0,
+            child_samples.size * state_dim * sizeof(tsk_bit_array_value_t));
+        for (k = 0; k < state_dim; k++) {
+            tsk_bit_array_get_row(
+                r_samples, (state_dim * (tsk_size_t) c) + k, &samples_row);
+            tsk_bit_array_get_row(&child_samples, k, &child_samples_row);
+            tsk_bit_array_add(&child_samples_row, &samples_row);
+        }
+        r_state->branch_len[c] = time[p] - time[c];
+        r_state->parent[c] = p;
+
+        in_parent = NULL;
+        while (p != TSK_NULL) {
+            for (k = 0; k < state_dim; k++) {
+                tsk_bit_array_get_row(
+                    r_samples, (state_dim * (tsk_size_t) p) + k, &samples_row);
+                tsk_bit_array_get_row(&child_samples, k, &child_samples_row);
+                tsk_bit_array_add(&samples_row, &child_samples_row);
+            }
+            compute_two_tree_branch_state_update(ts, c, in_parent, l_state, r_state,
+                state_dim, result_dim, +1, f, f_params, result);
+            in_parent = &child_samples;
+            c = p;
+            p = r_state->parent[p];
+        }
+    }
+out:
+    tsk_bit_array_free(&child_samples);
+    return ret;
+}
+
+static int
+tsk_treeseq_two_branch_count_stat(const tsk_treeseq_t *self, tsk_size_t state_dim,
+    const tsk_bit_array_t *sample_sets, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *TSK_UNUSED(norm_f),
+    tsk_size_t n_rows, const double *row_positions, tsk_size_t n_cols,
+    const double *col_positions, tsk_flags_t TSK_UNUSED(options), double *result)
+{
+    int ret = 0;
+    int r, c;
+    tsk_id_t *row_indexes, *col_indexes;
+    tsk_size_t i, j, k, row, col, *row_repeats, *col_repeats;
+    tsk_bit_array_t node_samples;
+    iter_state l_state, r_state;
```

Comment:
Whoops, need to initialise ``result_tmp=NULL`` or we'll have undefined behaviour on error conditions. Can you double check that all of the variables you are mallocing are initialised to NULL please?

Reply:
Ah right, I've fixed all of these now I think.